lib/Graph/NetworkComparator.cpp

```diff
@@ -0,0 +1,365 @@
+/**
+ * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Graph/NetworkComparator.h"
+
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Node.h"
+
+#include "glow/ExecutionEngine/ExecutionEngine.h"
+#include "glow/Graph/Hook.h"
+#include "glow/Graph/Utils.h"
+#include "glow/Support/Debug.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "verifier"
+
+using namespace glow;
+
+void NetworkComparatorBase::dumpTensors(
+    std::unordered_map<std::string, Tensor *> tensors, std::string layerName,
+    std::string prefix) {
+  // TODO: Need to add different flavours of dumping.
+  auto it = tensors.begin();
+  while (it != tensors.end()) {
+    std::error_code EC;
+    DCHECK(it->second) << "Tensor is Null\n";
+    llvm::raw_fd_ostream fs(prefix + "_" + it->first + "_" + layerName, EC);
+    it->second->dump(fs, std::numeric_limits<unsigned>::max());
+    fs.close();
+    it++;
+  }
+}
+bool NetworkComparatorBase::checkTensors(
+    std::unordered_map<std::string, Tensor *> &refOuts,
+    std::unordered_map<std::string, Tensor *> &checkOuts) {
+  if (refOuts.size() == checkOuts.size()) {
+    auto itRef = refOuts.begin();
+    auto itCheck = checkOuts.begin();
+    auto endRef = refOuts.end();
+    auto endCheck = checkOuts.end();
+    while (itRef != endRef && itCheck != endCheck) {
+      if (!itRef->second->isEqual(*(itCheck->second), numericCmpThreshold_,
+                                  true)) {
+        LOG(INFO) << "Error: " << itRef->first << "\n";
+        return false;
+      }
+      itRef++;
+      itCheck++;
+    }
+  } else {
```

Comment:
nit: Could check this first and return false early. Allows for the above logic in the current if case to be one nested level down.

Reply:
you mean something like:

```
  if (refOuts.size() != checkOuts.size()) {
    LOG(ERROR) << "Backends produced different number of results\n";
    return false;
  }

  auto itRef = refOuts.begin();
  auto itCheck = checkOuts.begin();
  auto endRef = refOuts.end();
  auto endCheck = checkOuts.end();
  while (itRef != endRef && itCheck != endCheck) {
    if (!itRef->second->isEqual(*(itCheck->second), numericCmpThreshold_,
                                /* verbose */ true)) {
      LOG(INFO) << "Error: " << itRef->first << "\n";
      return false;
    }
    itRef++;
    itCheck++;
  }
```

So the rest of the code is not indented in, right?